packages/bvaughn-architecture-demo/src/suspense/PointsCache.ts

```diff
@@ -12,89 +13,84 @@ import { isExecutionPointsLessThan } from "../utils/time";
-  }
-}
-
-async function fetchClosestPointForTime(
+async function fetchPointsBoundingTime(
   client: ReplayClientInterface,
   time: number,
-  record: Record<ExecutionPoint>,
-  wakeable: Wakeable<ExecutionPoint>
+  wakeable: Wakeable<ExecutionPoint>,
+  rethrowError: boolean
 ) {
   try {
-    const { point } = await client.getPointNearTime(time);
-
-    record.status = STATUS_RESOLVED;
-    record.value = point;
+    const pointsBoundingTime = await client.getPointsBoundingTime(time);
+    const point = getClosestPointInPointsBoundingTime(time, pointsBoundingTime);
 
+    // Pre-cache time-to-point match and insert into sorted ExecutionPoints array.
     preCacheExecutionPointForTime({ point, time });
 
+    // Also insert bounding points into sorted array; ignore duplicates!
+    if (sortedPointsBoundingTimes.length === 0) {
+      sortedPointsBoundingTimes.push(pointsBoundingTime);
+    } else {
+      const beforePoint = pointsBoundingTime.before.point;
+      const afterPoint = pointsBoundingTime.after.point;
+
+      let indexBegin = 0;
+      let indexEnd = sortedPointsBoundingTimes.length - 1;
+      let indexMiddle = 0;
+
+      while (indexBegin <= indexEnd) {
+        indexMiddle = Math.floor((indexBegin + indexEnd) / 2);
+
+        const currentPointsBoundingTime = sortedPointsBoundingTimes[indexMiddle];
+
+        if (
+          currentPointsBoundingTime.before.point === beforePoint &&
+          currentPointsBoundingTime.after.point === afterPoint
+        ) {
+          // Session.getPointsBoundingTime should never return overlapping ranges that are not exact matches.
+          // And we can skip adding add duplicate ranges.
+          // These shouldn't be common, but if two times within the same range are requested in parallel, we might see this.
+          indexMiddle = -1;
+          break;
+        } else if (currentPointsBoundingTime.after.point < beforePoint) {
+          indexBegin = indexMiddle + 1;
+        } else {
+          indexEnd = indexMiddle - 1;
+        }
+      }
+
+      if (indexMiddle >= 0) {
+        const closestPointsBoundingTime = sortedPointsBoundingTimes[indexMiddle];
+        if (pointsBoundingTime.before.point < closestPointsBoundingTime.after.point) {
+          sortedPointsBoundingTimes.splice(indexMiddle, 0, pointsBoundingTime);
+        } else {
+          sortedPointsBoundingTimes.splice(indexMiddle + 1, 0, pointsBoundingTime);
+        }
+      }
+    }
+
     wakeable.resolve(point);
   } catch (error) {
-    record.status = STATUS_REJECTED;
-    record.value = error;
+    if (rethrowError) {
+      throw error;
+    } else {
+      // TODO
+      // Note that we might be caching unloaded region errors by doing this.
+      // That's better than flooding the backend with repeated requests,
+      // but we also probably don't want to permanently cache this type of error.
+      // Maybe we should somehow make loaded regions part of the cache key?
+      timeToErrorMap.set(time, error);
```

Comment:
TBH I'm not sure what we should do here. The frontend could guard against asking for a point/time outside of the loaded range, so this should never _actually_ happen, but if we miss that guard...it's possible.